Json

eszterpos opened the pull request:
This branch contains the implementation of the handling of .json type extras strings.
It includes an implementation of a new data structure (listDict in the new tools/xtratools.py) that enables us to access the extra strings in a smoother way. The object stores a list of dictionaries 

In case the driver does not pass a JSON-formatted string, the mxtra dictionary creates an entry with {'info': string passed by the driver}.

Also, there is a new way of handling the output of extras, and to demonstrate this, I have slightly modified some existing examples. The novelty hides behind the new attribute of the trajectoryOutput object, 'xtratype' which enables the user to print out information to separate files depending on what extras strings have been received and what is of interest.

Moreover, I added a way to promote extras numerical data to be interpolated over in case the bead-contraction scheme is being used.

litman90 commented on examples/pes/instanton/ch4h-cbe-rate/instanton/20_non_uniform_discret/no_distr:
```diff
@@ -0,0 +1,21 @@
+#Bead    Energy (eV)
```
What is this file for?

litman90 commented on examples/pes/instanton/ch4h-cbe-rate/instanton/20_non_uniform_discret/input.xml:
```diff
@@ -8,7 +8,7 @@
    <system >
        <initialize nbeads='20'> 
-           <file mode='xyz' >  init.xyz </file>
+           <file mode='xyz' >  in.xyz </file>
```
why this file has been modified?
I see the init.xyz so this folder needs some cleaning.

litman90 commented on ipi/README:
```diff
@@ -0,0 +1,22 @@
+This branch contains the implementation of the .JSON extra files. 
```
This file should not be here

eszterpos replied:
deleted

litman90 commented on ipi/engine/motion/instanton.py:
```diff
@@ -938,6 +959,39 @@ def pre_step(self, step=None, adaptative=False):
         self.qtime = -time.time()
         info("\n Instanton optimization STEP {}".format(step), verbosity.low)
 
+        # Choosing the forcefield that has returned friction tensor in extras
```
The current PR is about the extras communication, I would delete this here.

eszterpos replied:
That is true, however, I think that the point why we are doing the whole modification of the extras input, is to have access to the friction. So I included this part to show how one accesses, the friction how to transfer that data structure into a friction object to work with.

mahrossi replied:
I agree that this can/should be deleted.

eszterpos replied:
I understand, yes :) I removed it.

litman90 replied:
is this resolved eszter? if so please press "resolve conversation"

litman90 commented on ipi/inputs/motion/instanton.py:
```diff
@@ -49,18 +49,6 @@ class InputInst(InputDictionary):
         )
     }
 
```
Why this inputAtributre has been eliminated?

eszterpos replied:
the whole block is completely repeated in the master branch, take a look :)

litman90 commented on ipi/inputs/outputs.py:
```diff
@@ -166,6 +166,22 @@ class InputTrajectory(InputValue):
+                "dipole",
+                "friction",
+                "hirshfeld",
+                "workfunction",
```
maybe change it to work_function to keep it consistent with atomic_stress?

eszterpos replied:
fair point! corrected it :)

litman90 commented on ipi_tests/regression_tests/tests/INSTANTON/200K/input.xml:
```diff
@@ -37,3 +37,4 @@
        </motion>
    </system>
 </simulation>
+
```
please delete extra space here

eszterpos replied:
deleted.

litman90 commented on ipi_tests/regression_tests/tests/GEOP/BFGS/pswater/RESTART:
```diff
@@ -0,0 +1,73 @@
+<simulation verbosity='medium' mode='static'>
```
Please delete this file

eszterpos replied:
deleted.

litman90 commented on .gitignore:
```diff
@@ -13,3 +13,4 @@ doc/manual.xml
 \#*
 *.orig
 regtest-ref
+JSONtest
```
The gitignore is not intended for specific tests purpose during development. Please delete this line

eszterpos replied:
yes, I had it there only not to synchronise the directory, I am going to make the regtests from the content of the above directory and will remove it from here, thanks! :)

litman90 replied:
is this resolved eszter? if so please press "resolve conversation"

litman90 commented on ipi/interfaces/sockets.py:
```diff
@@ -476,6 +486,55 @@ def dispatch(self, r):
         # marks the request as done as the very last thing
         r["status"] = "Done"
 
+    # def getjson(self):
```
please delete this

eszterpos replied:
deleted.

litman90 commented:
Please also add the regression tests required to test the new functionality

mahrossi commented:
Is this good to go? seems to me. Yair, approve?

eszterpos reviewed (commented):
> Please also add the regression tests required to test the new functionality

Yes, I added 2 tests, these are testing the xtratype functionality. I compared the numbers printed by the driver with the numbers that are being printed by i-pi.

litman90 commented:
Eszter could you explain how you generated the FRICTION_DIPOLE example? I don't get it.

litman90 commented:
I merged this branch with a copy of my working instanton branch to test it. 
The functionality works as expected =)

I have the  following suggestions,
 - add a default value to the new_x argument in 'transfer_forces_manual', so that function can be used without taking care of the presence of extras. A warning is in order though.
 - remove all the modifications to the instanton.py. This PR should not modify that file.
  (I already took all the relevant parts of the code to perform the tests ;) )

eszterpos commented:
There is a part in the instanton.py where we interpolate for the potential and the forces. I extended that part for the extras numerical inputs. I think this is a logical decision. Can you explain why should we remove that part? I agree that it would be more logical to be done in a different part of the code, so maybe placed somewhere else is better, could you maybe suggest a better place for this in the code?

litman90 commented:
> There is a part in the instanton.py where we interpolate for the potential and the forces. I extended that part for the extras numerical inputs. I think this is a logical decision. Can you explain why should we remove that part? I agree that it would be more logical to be done in a different part of the code, so maybe placed somewhere else is better, could you maybe suggest a better place for this in the code?

That part of the code doesn't exist anymore so it is not possible to say, put it here or there.
This PR makes available the "extra" information but it doesn't care how you use it in other parts of the code (like in this case, an interpolation)

eszterpos commented:
I added 2 regression tests testing the xtratools infrastructure, and removed the parts of the instanton.py file that are being partially reorganised and rewritten by @litman90 in the instanton 3.0 branch. I also added some fixes for the case when no extras are being passed. 

As the first regtests I created seemed to be unstable under the github CI, I now checked that the new ones are stable on both i-pi main repo and on the sabia repo, too. 

I consider the PR ready to merge from my side. What do you think, @litman90 and @mahrossi ? :)

litman90 commented on ipi/engine/forces.py:
```diff
@@ -894,35 +895,42 @@ def transfer_forces(self, refforce):
         assert len(self.mforces) == len(new_forces), msg
+        if new_x == None:
+            new_x = [[None] * self.nbeads] * len(self.mforces)
+            warning("WARNING: No extras information has been passed.")
```
This should not be a warning.

litman90 replied:
I will downgrade it to an info msg

litman90 commented:
It looks good to be merged after the tests finish without problems
